typescript/web/src/components/datasets/dataset-list.tsx

```diff
@@ -0,0 +1,121 @@
+import { gql, useQuery } from "@apollo/client";
+import { Flex } from "@chakra-ui/react";
+import type { Dataset as DatasetType } from "@labelflow/graphql-types";
+import {
+  PaginationProvider,
+  PaginationToolbar,
+  usePagination,
+} from "../pagination";
+import { Spinner } from "../spinner";
+import { DatasetCard } from "./dataset-card";
+import { DatasetCardBox } from "./dataset-card-box";
+import {
+  DatasetListProps,
+  DatasetListProvider,
+  useDatasetList,
+} from "./dataset-list.context";
+
+export const getDatasetsQuery = gql`
+  query getDatasets($where: DatasetWhereInput) {
+    datasets(where: $where) {
+      id
+    }
+  }
+`;
+const LoadingCard = () => (
+  <DatasetCardBox>
+    <Flex
+      w="100%"
+      h="2xs"
+      direction="column"
+      alignItems="center"
+      justify="center"
+    >
+      <Spinner color="brand.500" size="xl" />
+    </Flex>
+  </DatasetCardBox>
+);
+
+const Content = () => {
+  const {
+    loading,
+    datasets,
+    workspaceSlug,
+    setDeleteDatasetId,
+    setEditDatasetId,
+  } = useDatasetList();
+  return loading ? (
+    <LoadingCard />
+  ) : (
+    <>
+      {datasets.map(
+        ({
+          id,
+          slug,
+          images,
+          name,
+          imagesAggregates,
+          labelsAggregates,
+          labelClassesAggregates,
+        }) => (
+          <DatasetCard
+            key={id}
+            url={`/${workspaceSlug}/datasets/${slug}`}
+            imageUrl={images[0]?.thumbnail500Url ?? undefined}
+            datasetName={name}
+            imagesCount={imagesAggregates.totalCount}
+            labelClassesCount={labelClassesAggregates.totalCount}
+            labelsCount={labelsAggregates.totalCount}
+            editDataset={() => {
+              setEditDatasetId(id, "replaceIn");
+            }}
+            deleteDataset={() => {
+              setDeleteDatasetId(id, "replaceIn");
+            }}
+          />
+        )
+      )}
+    </>
+  );
+};
+
+const PaginationFooter = () => {
+  const { total } = usePagination();
+  return (
+    <>{total > 0 && <PaginationToolbar pos="fixed" bottom={0} w="100%" />}</>
+  );
+};
+
+export const DatasetList = (props: DatasetListProps) => {
+  const { workspaceSlug } = props;
+  const { data: datasetsResult, loading: totalCountLoading } = useQuery<{
```

Comment:
Shouldn't it be in the DatasetListProvider (which could include the `PaginationProvider`)?

Reply:
This query is used before mounting `DatasetListProvider`. The return is passed to `PaginationProvider` and used then for determine if more data has to be loaded.

As `DatasetListProvider` depends on data from `PaginationProvider`, no it shouldn't be the way you ask in my opinion 😉